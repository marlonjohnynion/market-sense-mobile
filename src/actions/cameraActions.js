import { ImagePicker } from 'expo'
import firebase from '../common/firebase'
import { ActionSheet } from 'native-base'
import { toast } from '../common/helpers'

export const pickImage = () => {
  return async (dispatch) => {
    const pickedImage = await ImagePicker.launchImageLibraryAsync({
      base64: true,
      allowsEditing: false,
      aspect: [4, 3]
    })
    dispatch({ type: 'IMAGE_CHOSEN', image: pickedImage })
  }
}

export const takeImage = () => {
  return async (dispatch) => {
    const takenImage = await ImagePicker.launchCameraAsync({
      base64: true,
      allowsEditing: false,
      aspect: [4, 3]
    })
    dispatch({ type: 'IMAGE_CHOSEN', image: takenImage })
  }
}

export const getFirebaseImageUrl = async (pickedImage) => {
  try {
    const imageUri = pickedImage.uri
    const imageTitle = imageUri.split('/').pop()
    const image = await fetch(imageUri)
    const imageBlob = await image.blob()
    const storageRef = await firebase.storage().ref('/products')
    const imageRef = storageRef.child(imageTitle)
    const result = await imageRef.put(imageBlob)
    return result.downloadURL
  } catch (e) {
    toast(e)
  }
}

export const chooseImageOrigin = () => {
  return async (dispatch) => {
    const actions = ['Take Picture', 'Choose Image', 'Cancel']
    const cancelIndex = 2
    ActionSheet.show(
      {
        options: actions,
        cancelButtonIndex: cancelIndex,
        title: 'Upload Picture'
      },
      buttonIndex => {
        const status = actions[buttonIndex]
        if (status === actions[0]) {
          dispatch(takeImage())
        } else if (status === actions[1]) {
          dispatch(pickImage())
        }
      }
    )
  }
}