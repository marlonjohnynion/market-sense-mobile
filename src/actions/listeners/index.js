import { realtimeOrderListeners } from './orderListeners'
import { realtimeProductListeners } from './productListeners'

export const initializeListeners = (dispatch, state) => {
  realtimeOrderListeners(dispatch, state)
  realtimeProductListeners(dispatch, state)
}
