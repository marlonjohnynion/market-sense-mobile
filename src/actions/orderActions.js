import firebase from '../common/firebase'
import { generateInvoiceNumber } from '../common/helpers'
import { ActionSheet } from 'native-base'

export const addOrder = (order) => {
  return async (dispatch, getState) => {
    const { selectedProduct } = getState().products
    const parsedProductData = {
      key: selectedProduct.key,
      productPrice: parseFloat(selectedProduct.productPrice),
      lotSize: parseFloat(selectedProduct.lotSize)
    }
    const payload = {
      ...order,
      sellerKey: selectedProduct.ownerKey,
      invoiceNumber: generateInvoiceNumber(),
      quantity: parseInt(order.quantity),
      status: 'Pending',
      userKey: getState().user.uid,
      product: {
        ...parsedProductData
      }
    }
    const newOrder = {
      ...payload,
      product: {
        ...selectedProduct
      }
    }
    try {
      await firebase.database().ref('/orders').push(payload)
      dispatch({ type: 'ADD_ORDER', order: newOrder })
    } catch (e) {
      throw new Error(e)
    }
  }
}

export const loadOrders = () => {
  return async (dispatch) => {
    try {
      const orders = []
      const data = await firebase.database().ref('/orders').once('value')
      for (const [key, resData] of Object.entries(data.val())) {
        const order = {
          key: key,
          ...resData
        }
        const { product } = order
        const productData = await firebase.database().ref(`/products/${product.key}`).once('value')
        order.product = {
          ...productData.val(),
          ...order.product
        }
        orders.push(order)
      }
      dispatch({ type: 'LOAD_ORDERS', orders: orders })
    } catch (e) {
      throw new Error(e)
    }
  }
}

export const loadOrdersMadeToUser = () => {
  return async (dispatch, getState) => {
    const userKey = getState().user.uid
    try {
      const orders = []
      let data = await firebase.database().ref('/orders')
      data = await data.orderByChild('sellerKey').equalTo('Bkca5APqUbTcHFRjuqStnu9Mlph1').once('value')
      console.log(data)
      for (const [key, resData] of Object.entries(data.val())) {
        const order = {
          key: key,
          ...resData
        }
        const { product } = order
        const productData = await firebase.database().ref(`/products/${product.key}`).once('value')
        order.product = {
          ...productData.val(),
          ...order.product
        }
        orders.push(order)
      }
      dispatch({ type: 'LOAD_ORDERS_MADE_TO_USER', orders: orders })
    } catch (e) {
      throw new Error(e)
    }
  }
}

export const viewOrderReceipt = (order) => ({
  type: 'VIEW_ORDER_RECEIPT',
  order: order
})

export const viewSaleReceipt = (order) => ({
  type: 'VIEW_SALE_RECEIPT',
  order: order
})

export const updateOrderStatus = (order) => {
  return async (dispatch) => {
    const actions = ['Delivered', 'Received', 'Pending', 'Cancelled', 'Cancel Status Update']
    const destructiveIndex = 3
    const cancelIndex = 4
    ActionSheet.show(
      {
        options: actions,
        cancelButtonIndex: cancelIndex,
        destructiveButtonIndex: destructiveIndex,
        title: 'Update Status'
      },
      buttonIndex => {
        const status = actions[buttonIndex]
        if (status !== actions[cancelIndex]) {
          dispatch(sendStatusUpdate(order, status))
        }
      }
    )
  }
}

export const sendStatusUpdate = (order, status) => {
  return async (dispatch) => {
    try {
      const orderRef = firebase.database().ref('/orders')
      await orderRef.child(order.key).update({
        status: status
      })
    } catch (e) {
      console.log(e)
    }
  }
}
