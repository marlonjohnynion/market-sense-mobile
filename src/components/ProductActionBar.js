import React from 'react'
import { StyleSheet, View, Text } from 'react-native'
import { Grid, Col, Button } from 'native-base'

const ProductActionBar = ({product}) => {
  return (
    <View style={styles.productBottomBar}>
      <Grid>
        <Col style={styles.left}>
          <ProductPrice price={product.productPrice} />
          <Text>Free Delivery</Text>
        </Col>
        <Col style={styles.right}>
          <ActionButton text={'CHECK AVAILABILITY'}/>
        </Col>
      </Grid>
    </View>
  )
}

export const ProductPrice = ({price}) => {
  return (
    <Text style={styles.productPrice}>
      {price}
    </Text>
  )
}

export const ProductRating = ({rating}) => {
  return (
    <Text style={styles.productRating}>
      {rating}
    </Text>
  )
}

export const ActionButton = ({text}) => {
  return (
    <Button block success>
      <Text>{text}</Text>
    </Button>
  )
}

const styles = StyleSheet.create({
  productPrice: {fontSize: 16, paddingTop: 6, fontWeight: '800'},
  productRating: {fontSize: 16, fontWeight: '800', color: 'green'},
  productBottomBar: {
    height: 75,
    paddingTop: 15,
    paddingRight: 20,
    paddingLeft: 20,
    backgroundColor: '#fff',
    borderColor: 'black',
    borderTopWidth: 0.3,
    marginTop: 20
  },
  right: {flex: 0.6},
  left: {flex: 0.6}
})

export default ProductActionBar
