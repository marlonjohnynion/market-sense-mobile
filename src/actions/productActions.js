import firebase from '../common/firebase'
import { toast, toastGenericErrorMsg, isNumber } from '../common/helpers'
import { getFirebaseImageUrl } from './cameraActions'
import { startLoading, endLoading } from './creators/app'
import { clearTakenImage } from './creators/image'
import {
  changeSearchTerm,
  productFilter,
  viewProductAvailability
} from './creators/product'

const validateProductBeforeSubmission = (product) => {
  const {
    productTitle,
    productCategory,
    productDescription,
    productLocation,
    lotSize,
    maxLots,
    productPrice,
    minDeliveryDate,
    maxDeliveryDate,
    productImageUri
  } = product
  if (!productTitle || !productCategory || !productDescription || !productLocation || !lotSize ||
    !maxLots || !productPrice || !minDeliveryDate || !maxDeliveryDate) {
    throw new Error('Please fill up all the fields!')
  }
  if (new Date(minDeliveryDate) > new Date(maxDeliveryDate)) {
    throw new Error('Min delivery date is greater than max delivery date!')
  }
  if (!isNumber(maxLots) || parseInt(maxLots) <= 0) {
    throw new Error('Max lot should be a valid number!')
  }
  if (!isNumber(lotSize) || parseInt(lotSize) <= 0) {
    throw new Error('Lot size should be a valid number!')
  }
  if (!isNumber(productPrice) || parseInt(productPrice) <= 0) {
    throw new Error('Product price should be a valid number!')
  }
  if (!productImageUri) {
    throw new Error('Please select a product image!')
  }
}

export const addProduct = (product) => {
  return async (dispatch, getState) => {
    try {
      dispatch(startLoading())
      product.productImageUri = await getFirebaseImageUrl(getState().image.chosenImage)
      validateProductBeforeSubmission(product)
      const payload = {
        ownerKey: getState().user.uid,
        ...product
      }
      await firebase.database().ref('/products').push(payload)
      dispatch(clearTakenImage())
      dispatch(endLoading())
      toast('Product has been listed!')
    } catch (e) {
      toast(e)
    }
  }
}

export const updateProduct = (product) => {
  return async (dispatch, getState) => {
    try {
      const newProductImage = getState().image.chosenImage
      if (newProductImage) {
        product.productImageUri = await getFirebaseImageUrl(newProductImage)
      }
      validateProductBeforeSubmission(product)
      const payload = { ...product }
      delete payload.key
      delete payload.productOwner
      delete payload.ownerAvatar
      const productRef = await firebase.database().ref('/products')
      await productRef.child(product.key).update(payload)
      dispatch(clearTakenImage())
      toast('Product has been updated!')
    } catch (e) {
      toast(e)
    }
  }
}

export const searchTermChange = (value) => {
  return (dispatch) => {
    dispatch(changeSearchTerm(value))
  }
}

export const deleteProduct = (product) => {
  return async () => {
    try {
      const productRef = await firebase.database().ref('/products')
      await productRef.child(product.key).remove()
    } catch (e) {
      toastGenericErrorMsg()
    }
  }
}

export const checkProductAvailability = () => {
  return async (dispatch) => {
    dispatch(viewProductAvailability())
  }
}

export const filterProducts = () => {
  return (dispatch, getState) => {
    dispatch(productFilter(getState().products.searchTerm))
  }
}
