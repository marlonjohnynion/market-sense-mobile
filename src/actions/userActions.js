import * as actions from './types/userActionTypes'
import firebase from '../common/firebase'
import { toast } from '../common/helpers'
import { initializeListeners } from './listeners'

export const loginSuccess = (values) => ({
  type: actions.LOGIN,
  email: values.email,
  uid: values.uid
})

export const loginFail = (error) => ({
  type: actions.LOGIN_FAILED,
  error: error
})

export const registerFail = (error) => ({
  type: actions.REGISTER_USER_FAILED,
  error: error
})

export const registerUser = (email, password) => {
  return async (dispatch) => {
    try {
      await firebase
        .auth()
        .createUserWithEmailAndPassword(email, password)
      // if user is registered, we will login automatically
      if (firebase.auth().currentUser) {
        dispatch(loginSuccess(email, password))
      }
    } catch (e) {
      dispatch(registerFail(e))
    }
  }
}

export const authenticateUser = (values) => {
  return async (dispatch, getState) => {
    const { email, password } = values
    try {
      await firebase
        .auth()
        .signInAndRetrieveDataWithEmailAndPassword(email, password)
      if (firebase.auth().currentUser) {
        values.uid = firebase.auth().currentUser.uid
        dispatch(loginSuccess(values))
        initializeListeners(dispatch, getState())
      }
    } catch (e) {
      toast('Wrong password/email combination')
    }
  }
}
