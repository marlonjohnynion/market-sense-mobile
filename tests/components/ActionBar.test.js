import React from 'react'
import { shallow } from 'enzyme'
import ProductActionBar, {ProductPrice, ActionButton} from '../../src/components/ProductActionBar'
import { checkProductAvailability } from '../../src/actions/productActions'

const product = {
  productImageUri: 'https://firebasestorage.googleapis.com/v0/b/marketsense-b09f8.appspot.com/o/products%2Fp0.png?alt=media&token=c305d51d-1b8b-4fbb-bd10-0c03a5c5805d',
  productTitle: 'Carrots',
  productCategory: 'Vegetable',
  productOwner: 'Marlon John Ynion',
  productDescription: 'Some vegetables',
  productPrice: 5,
  productRating: 5
}

const actions = {
  checkProductAvailability: checkProductAvailability
}

describe('<ProductActionBar />', () => {
  let wrapper

  beforeEach(() => {
    wrapper = shallow(<ProductActionBar product={product} actions={actions}/>)
  })

  it('renders the action bar', () => {
    expect(wrapper).toHaveLength(1)
  })

  it('renders the action button', () => {
    expect(wrapper.find(ActionButton)).toHaveLength(1)
  })

  it('renders the product price', () => {
    expect(wrapper.find(ProductPrice)).toHaveLength(1)
  })

  it('renders the component as expected', () => {
    expect(wrapper).toMatchSnapshot()
  })
})
