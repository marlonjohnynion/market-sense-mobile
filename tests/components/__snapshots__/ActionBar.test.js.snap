// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProductActionBar /> renders the component as expected 1`] = `undefined`;
